modules/rgbd/src/colored_kinfu.cpp

```diff
@@ -272,29 +273,31 @@ bool ColoredKinFuImpl<MatType>::updateT(const MatType& _depth, const MatType& _r
     else
         rgb = _rgb;
 
-    cv::Ptr<OdometryFrame> newFrame = icp->makeOdometryFrame(rgb, depth, noArray());
-    //TODO: fix it
-    // This workaround is needed because we want to keep color image in newFrame
-    // FastICP doesn't use color info and doesn't prepare its pyramids
-    newFrame->setPyramidLevels(params.icpIterations.size());
+    OdometryFrame newFrame = icp.createOdometryFrame();
 
-    icp->prepareFrameCache(newFrame, OdometryFrame::CACHE_SRC);
+    newFrame.setImage(rgb);
+    newFrame.setDepth(depth);
 
     if(frameCounter == 0)
     {
+        icp.prepareFrames(newFrame, newFrame);
+
         // use depth instead of distance
         volume->integrate(depth, rgb, params.depthFactor, pose, params.intr, params.rgb_intr);
-        newFrame->setPyramidAt(rgb, OdometryFrame::PYR_IMAGE, 0);
+        newFrame.setPyramidLevel(params.icpIterations.size(), OdometryFramePyramidType::PYR_IMAGE);
```

Comment:
I guess `setPyramidLevel` is called because all pyramid vectors are empty by default.
I left a comment in main repo's PR on that.